### PR TITLE
add support for rpm %attr for files

### PR DESCRIPTION
**Pull request by olisol1**
instead of relying on chmod/chown for files use the %attr directive of rpm. Particularly useful for files_extras

Needed this for a project of mine, so I might as well give back to the community.

### CODE REVIEW

**Comment by kevinconway** [comment]
Thanks for the contribution! I appreciate you making a pull request. Looking at the PR, it seems like this would be backwards compatible so I don't have any issues merging it. For my own education, what issue did you find with the `chmod` that caused you to switch it to `attr`?

The tests are failing but it's unrelated to your changes. This happens regularly for a stable project like this that doesn't change much. The world changes around the project and the builds start to break. I see that you pushed another patch to fix the enchant installation. I really appreciate that and didn't expect it. I meant to say something after I enabled the tests but I started having internet issues. I'll fix the build process after I merge your code. It looks like it's an issue with trying to build on 3.6 which is EOL as far as I know.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -45,11 +45,11 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        pyver: ["3.6", "3.7", "3.8", "3.9", "3.10"]
+        pyver: ["3.6.15", "3.7.15", "3.8.16", "3.9.16", "3.10.10"]
```
Comment (kevinconway) [inline .github/workflows/ci.yml:48]:
```suggestion
        pyver: ["3.7", "3.8", "3.9", "3.10", "3.11"]
```

I think this is the list of currently supported Python versions.

**Comment by kevinconway** [comment]
@olisol1 Thanks again for the PR. I'll sort out the build issue before I create a new release.

For the release notes, could you tell me what `attr` solved for you?

**Comment by olisol1** [comment]
> Thanks for the contribution! I appreciate you making a pull request. Looking at the PR, it seems like this would be backwards compatible so I don't have any issues merging it. For my own education, what issue did you find with the `chmod` that caused you to switch it to `attr`?

Thank you for having this project it's a lifesaver!
They are basically equivalent but you can do it on a file basis in a `%file` section, so it is more flexible for example if you want to restrict the use of some file after install, but do not care in dev environment. (in my case i had to give a specific group to a few files in my package)
The `chmod` was done in `%post`, thus to be able to use `%attr` directives and them not be overwritten, I had to remove the chmods. Also using `chmod` was redundant with the `%defattr` directive provided by the file_permissions extension  which already defines the files properties to the default user:group when `%attr` isn't used.

> The tests are failing but it's unrelated to your changes. This happens regularly for a stable project like this that doesn't change much. The world changes around the project and the builds start to break. I see that you pushed another patch to fix the enchant installation. I really appreciate that and didn't expect it. I meant to say something after I enabled the tests but I started having internet issues. I'll fix the build process after I merge your code. It looks like it's an issue with trying to build on 3.6 which is EOL as far as I know.

Ah I see. I'm still a novice in git-hub opensource work, so I meant to do my best :)

**Comment by kevinconway** [comment]
Ah, ok. So it allowed you to set permissions on individual files instead of applying a single set of permissions to the entire set of files.

I see now that there is a small backwards compatibility issue in the PR. The very original version of the `files_extras` plugin was much simpler than today's. It allowed only selecting files with strings like `/source/file:/destination/file`. The `files_extras` plugin requires the `files_permissions` plugin because that was, originally, the only way to communicate the user and group that should be used for the files. That's when I added the post hook to run `chmod`. Later, another contributor extended the `files_extras` configuration to accept a dictionary so they could add more complex instructions like annotating files as documentation or configuration in the RPM spec. However, we kept the requirement for `files_permissions` in place because there was still no way to specify permissions otherwise.

Your change adds the ability to set permissions for specific files but at the cost of discarding the `files_permissions` configuration. That is, users _must_ now specify all file permissions using the new configuration values because the old ones from `files_permissions` are ignored. What I mean by that is this block of code that was removed:
```python
spec.blocks.post.append(
    'chown -R '
    '%{{file_permissions_user}}:%{{file_permissions_group}}  '
    '/{0}'.format(file_.dest)
 )
```
The `file_permissions_user` and `file_perimssions_group` values are RPM spec macros that are defined by the `file_permissions` plugin. Existing users of rpmvenv will expect that those permissions are used for all extra files by default because that's how it's worked in the past.

I think the easiest way to resolve this is to set the default values of the `%attr` to the old marcos and only override them when the new configuration values are present. For example:
```python
for file_ in config.file_extras.files:

    file_directive = '%attr(-, %{file_permissions_user}, %{file_permissions_group}) '
    if file_.file_attr is not None:
        # file with attributes modifiers (permissions, user, group)
        file_directive = '%attr({0}, {1}, {2}) '.format(
                file_.file_attr['permissions'],
                file_.file_attr['user'],
                file_.file_attr['group'],
            )
```

This way, the macro is used by default and all existing user expectations are preserved. At the same time, any users of the new options you've added will get the new behavior but only for the specific files they've configured to use different permissions.

If you have the time and motivation to make this change and test it out for your use case then please do and send me another PR. If not then I'll try to look into it myself when I have time but that might be a while. I'll have to hold off on a new release until this can be made backwards compatible so that folks using this for builds don't start having issues.

**Comment by kevinconway** [comment]
@olisol1 I had another chance to read this over and I realized I was wrong. The `file_permissions` plugin sets a `%defattr` directive that should set the default permissions for all files that are installed. With your patch, it would still only be files with an explicit `%attr` that are different than the default. So it's not a breaking change.

I went back and looked at the code history. I'm fairly confident now that the `chmod` line was only there because the original version of `files_extras` did all file operations with shell scripts in `post`. At a certain point in time, I added the ability to use other kinds of directives, like `files`, to the RPM spec. I even added the `%defattr` but never removed the `chmod` that it made redundant.

So this should be fine to release once I get another patch in. I found one thing we missed in the PR: https://github.com/kevinconway/rpmvenv/pull/117/files#diff-bf2d6a2ff80506b20ac1203a543ed2e231bf9cf2132a0e96a7f393155723d34eR65 . The tests didn't catch it because they don't include a case for having both the file attributes and a file type set. When they are both set then it would duplicate the attribute like `%attr(...) %attr(...) /file/path`. I'll fix it and get a new release out so you can use your changes.

**Comment by olisol1** [comment]
Sorry I did not manage to respond yesterday and I see you already fixed everything.
Thanks you very much for the speedy support!

**Pull request merged**